selfdrive/car/ford/carstate.py

```diff
@@ -29,9 +29,14 @@ def update(self, cp, cp_cam):
     self.unsupported_platform = (cp.vl["VehicleOperatingModes"]["TrnAinTq_D_Qf"] == 0 and
                                  self.CP.carFingerprint not in CANFD_CAR)
 
-    # Occasionally on startup, the ABS module recalibrates the steering pinion offset, so we need to block engagement
-    # The vehicle usually recovers out of this state within a minute of normal driving
-    self.vehicle_sensors_valid = cp.vl["SteeringPinion_Data"]["StePinCompAnEst_D_Qf"] == 3
+    if self.CP.carFingerprint in CAN_EDGE:
+      # Ford Edge calibrated steering wheel angle comes from the IPMA_ADAS instead of PSCM
+      # Check for Ford Edge Unknown/Invalid steering wheel position
+      self.vehicle_sensors_valid = cp.vl["ParkAid_Data"]["ExtSteeringAngleReq2"] < 32766
```

Comment:
If `SteeringPinion_Data` on this car, but is just on another ID (and that ID doesn't conflict with any existing message), we can add it to the DBC with a prefix such as `Alt`, and detect the absence 0x7e to use the new alternate message. This is common for other brands, such as Hyundai where we support many different configurations.

Reply:
I can update the existing ford_lincoln_base_pt DBC with the alternate ID for documentation/future-use purposes. However, `SteeringPinion_Data: StePinCompAnEst_D_Qf` and `SteeringPinion_Data: StePinComp_An_Est` signals on the Edge are not used/available. They remain at 0:Faulty/-1600 deg, even when remapped to the new ID.

![image](https://github.com/commaai/openpilot/assets/735082/5f01395d-b5ea-49dc-896c-bc71abde326e)

I initially tried using `SteeringPinion_Data: StePinRelInit_An_Sns`, but this signal is not calibrated 0=center. The offset changed every time the vehicle was started. OP would only function if the offset was within `paramsd.OFFSET_MAX`, which was very unreliable.

I found the signal `ParkAid_Data: ExtSteeringAngleReq2` contains calibrated steering wheel angle on the Edge. I surveyed other drivers on the #ford Discord channel and no other models have any data at this signal. This leads me to believe the Edge has moved the calibrated steering wheel angle to the IPMA and it is not available in the PSCM.

![image](https://github.com/commaai/openpilot/assets/735082/c81c76e6-7cbd-4f5b-a3dc-061d43be56a1)

Reply:
Dang, that's unfortunate. What about `BrakeSnData_6->StePinOffst_An_Est`?

Reply:
On our Bronco, when the Qf signal is invalid, `StePinComp_An_Est` is 0 degrees (0b011111010010000), but yours is just zero'd out as bytes (hence the -1600), that is a meaningful difference, and perhaps tells us that it isn't a signal that will ever be filled out, so we can ignore using that.

Reply:
No data at `BrakeSnData_6`

![image](https://github.com/commaai/openpilot/assets/735082/8736c199-aaf1-4931-a358-34c7ab37463e)

-

Are you asking about getting a wrong angle offset at `SteeringPinion_Data->StePinRelInit_An_Sns`? Reviewing a couple of routes, it looks like it's initialized to 0 at ignition so the offset is equal to steering wheel angle at ignition

![image](https://github.com/commaai/openpilot/assets/735082/e40a040f-7aa6-43a8-aef5-2479d7417da9)
![image](https://github.com/commaai/openpilot/assets/735082/6c834d09-74ae-4e50-9c01-77e5e2592011)
![image](https://github.com/commaai/openpilot/assets/735082/00c07791-1278-408f-bca7-65a3b73d5408)

-

The responses I mentioned receiving earlier from other Ford drivers on #ford Discord all seemed to be CAN_FD vehicles. I just found another route from a CAN vehicle, `ParkAid_Data->ExtSteeringAngleReq2` was identical to `SteeringPinion_Data->StePinComp_An_Est`. Does your Bronco have data at `ParkAid_Data->ExtSteeringAngleReq2`? Should this be the steering wheel angle for all CAN vehicles rather than just EDGE?

Reply:
Here's the route you asked for. I toggled the vehicle on and off 50 times and then left it on for about a minute to collect data. Accessory power only. The steering wheel is turned 90 to the left, so should read about +90 when calibrated
3d4e31f900dd3a88|2023-12-20--00-45-19

Reply:
> * [ ]  Still should check that it never is not set across lots of Edge routes

Wrote a script to analyze the rlogs (or qlogs if rlogs not uploaded) for `StePinComp_An_Est` data for a list of routes. I ran this on about 500 miles worth of routes from my Edge. I also threw a couple of other Fords in as a control. Please let me know if you want to see anything more

[check_steering_data.py.zip](https://github.com/commaai/openpilot/files/13882429/check_steering_data.py.zip)
![image](https://github.com/commaai/openpilot/assets/735082/6dc0598a-d522-4cc6-8521-e1718f0055c8)